### PR TITLE
Restore support for native_c_compiler and bytecomp_c_compiler

### PR DESCRIPTION
**Pull request by yallop**
Alongside several improvements to the build system, PR #1114 introduced a couple of breaking changes in the output of `ocamlc -config`: the variables `bytecomp_c_compiler` and `native_c_compiler` were removed, perhaps inadvertently.

The removed variables have been [part of the interface since OCaml 3.09 (2005)](https://github.com/ocaml/ocaml/commit/6c9bac39d4), and several packages rely on them.

This PR reconstructs the values of `bytecomp_c_compiler` and `native_c_compiler` from the new variables `c_compiler`, `ocamlc_cflags`, `ocamlc_cppflags`, etc.,  restoring backwards compatibility.

There's some discussion of the change under [Mantis PR 7650](https://caml.inria.fr/mantis/view.php?id=7650).

### CODE REVIEW

**Comment by dra27** [comment]
I'd been under the mistaken impression earlier that this change had been made in 4.05.0 (it's easy, for me at least, to forget which build system improvements have been in which particular branch).

That we haven't released it yet changes my view on whether we should do it (i.e. I agree we should). One other thing I noticed is that the new ocamlc_cppflags and ocamlopt_cppflags are not included in the Config.print_config, and I think that must be an omission too (@shindere?)

**Comment by shindere** [comment]
Would it be okay to not declare the varialbes in the .mli file so that
people are not encouraged to use them?

It is worth noting that the bytecode_c_compiler and native_c_compiler
names were introduced for historical reasons but are at the same time
misleading.

Indeed, if I remember correctly what @xavierleroy explained (can you
confirm or correct me, @xavierleroy, please?), historically there was a
need to use two different C compilers to compile OCaml's runtime because
one compiler was more strict but produced less efficient code while the
other was the other way around. So one compiler was more interesting to
produce a runtime that would be used to interprete bytecode, while the
other C compiler produced a runtime which was more convenient to support
native code.

So all this has not much to do with whether the compiled C code will be
used with either bytecode-compiled or native OCaml programs.

To sum up, sorry to have introduced this backward-incompatibility, I'm
very fine with restoring the former varialbes to keep this backward
compatibility.

At the same time, I think it would be great if the packages could stop
using these variables (perhaps by testing whether they are present or
something like that) so that they can ultimately be removed completely.

**Comment by yallop** [comment]
> Would it be okay to not declare the varialbes in the .mli file so that people are not encouraged to use them?

That's fine with me.  I've pushed a commit that removes them from the `.mli`.

> At the same time, I think it would be great if the packages could stop using these variables (perhaps by testing whether they are present or something like that) so that they can ultimately be removed completely.

It would be useful to have a deprecation mechanism here.  Having warnings logged to the console (as `@deprecated` does for functions) is a useful way to communicate such changes and a good incentive for maintainers to switch to newer mechanisms.

Here's one possible approach: extend `-config` so that it optionally takes an argument specifying the config variable to print, and then print a warning to stderr when a deprecated variable is requested, like this:

```
$ ocamlc -config ext_obj
.o
$ ocamlc -config native_c_compiler
gcc -Wall -D_FILE_OFFSET_BITS=64 -D_REENTRANT
Warning: native_c_compiler is deprecated             # printed to stderr
         Please use 'c_compiler', 'ocamlopt_cflags'  # printed to stderr
         and 'ocamlopt_cppflags' instead             # printed to stderr
```

This approach would have the additional advantage that there would no longer be a need to parse the output of `ocaml -config` with external tools like `sed` or `awk`.

**Comment by dra27** [comment]
I'm strongly in favour of adding the extra parameter to `-config` - that's been on my back-burner list for a while!

Given the tight release cycle we're aiming for 4.06, that'd probably have to wait for 4.07, though.

**Comment by gasche** [comment]
As mentioned elsewhere, I support having `-config-var var` (please don't reuse `-config`), which would make scripting easier by printing the value with a success return code when the key exists, and print nothin with an error code when it does not. (That's matter for a separate PR, of course.)

**Review by gasche — APPROVED** [review]

**Comment by shindere** [comment]
Thanks for the commit.

We indeed discussed the possibility to introduce a way to deprecate
these variables. This would be useful in general.

Regarding the implementation you suggest, my understanding is that the
tools provided by OCaml to parse the command-line are not so comfortable
with a command-line option taking an optional argument, or am I wrong
here?

So such a mechanism that only prints one variable might need to rely on
another command-line option, different from -config.

And of course, is such a mechanism would be introduced, the existing
scripts would have to be modified to make use of it. As long as this is
not done, they won't even see the deprecation information.

This is not to say what you propose shouldn't be done. Just thinking
aloud, here.

**Comment by shindere** [comment]
@yallop could you please break the
long-lines in utils/config.mlp ?

Although there is not official deprecaton mechanism, it may be worth
at least marking each of these variables as deprecated with a comment
close to its definition in utils/config.mlp

**Comment by yallop** [comment]
@shindere:
> could you please break the long-lines in utils/config.mlp ?

Yes; I've done this.

> it may be worth at least marking each of these variables as deprecated with a comment close to its definition in utils/config.mlp

I've added a comment, as suggested, and squashed the commits.

> So such a mechanism that only prints one variable might need to rely on another command-line option, different from -config.

Indeed -- @gasche's suggestion (a separate `-config-var` option that always takes an argument) sounds like a better design.

**Comment by shindere** [comment]
yallop (2017/10/04 12:08 +0000):
> @shindere:
> > could you please break the long-lines in utils/config.mlp ?
> 
> Yes; I've done this.

Thanks.

> > it may be worth at least marking each of these variables as deprecated with a comment close to its definition in utils/config.mlp
> 
> I've added a comment, as suggested, and squashed the commits.

Thanks a lot for that, too.

Two more comments, but feel free to ignore them if you are bored. :)

1. I would have made it clearer in the comment that these variables will
really be removed.

2. (completely subject to discussion, or to be completely ignored). I'm
not a big fan of the "no change entry needed" in the commit messages
themselves. I tend to prefer to have this as a PR label. But perhaps
it's just a matter of personal taste.

> > So such a mechanism that only prints one variable might need to rely on another command-line option, different from -config.
> 
> Indeed -- @gasche's suggestion (a separate `-config-var` option that
> always takes an argument) sounds like a better design.

Agreed.

To me the PR is okay to be merged, even if you prefer not to take the
comments above into account, @yallop.

**Comment by yallop** [comment]
> 2. (completely subject to discussion, or to be completely ignored). I'm not a big fan of the "no change entry needed" in the commit messages themselves. I tend to prefer to have this as a PR label.

I think it's often better to use PR labels, too.  The commit message doesn't seem like the ideal place for that particular piece of metadata.  I used the commit message in this PR for entirely pragmatic reasons.  Using PR labels needs a maintainer to check whether a Changes entry is needed, add a label, then restart the Travis build.  In contrast, adding a line to a commit message is easy to do myself, so it's a quicker route to a passing status indicator on the PR.

**Comment by shindere** [comment]
I understand. I was not aware it was necessary to be a maintainer to use
labels, sorry.

**Comment by gasche** [comment]
Merged, cherry-picked in 4.06 ( d9c343fb516c6d974ad5df6e7e0fb8f03f68a461 ). Because we need to have `ctypes` in the testing loop, I may try to get another beta released soon-ish. (The other option is to just test from the 4.06+trunk switch, which is also reasonable.)

**Comment by gasche** [comment]
@yallop, could you confirm that, except for this regression, there exists a recent released version of ctypes that works under 4.06? I'm considering doing another beta release, but there is no point if ctypes still need more fixes.

**Comment by yallop** [comment]
@gasche: the latest ctypes release, 0.12.1 (and probably several earlier releases) install cleanly with 4.06 trunk, and I expect that they also work with the latest 4.06.0 beta.  The ctypes 0.12.1 tests, however, don't build with 4.06 trunk due to the `native_c_compiler` issue and a safe-string problem.  I'm planning to make a new ctypes release soon with tests that pass with OCaml 4.06.

**Comment by gasche** [comment]
Thanks!

**Comment by gasche** [comment]
If we wish to deprecate the `{bytecode,native}_c_compiler` configuration lines, should we not also include `ocaml{c,opt}_cppflags` lines? It seems that currently it is not possible to get the same information as `bytecomp_c_compiler` from the other variables, and @yallop explained that the cppflags are important to his use-case -- he needs the `_FILE_OFFSET_BITS` etc. to match.

**Comment by dra27** [comment]
Indeed - I'd noted above that they are missing. I think it's just an oversight, and one which we should fix (separately) for 4.06
